pkg/processor/procbuiltin/unwrap.go

```diff
@@ -93,6 +96,178 @@ func (p *unwrapProcessor) Process(_ context.Context, in record.Record) (record.R
+			metadata[k] = fmt.Sprint(v)
+		}
+	default:
+		return metadata, cerrors.Errorf("expected a record.Metadata or a map[string]interface{}, got %T", m)
+	}
+	return metadata, nil
+}
+
+// UnwrapKey extracts key from a structuredData record.
+func (o *openCDCUnwrapper) UnwrapKey(structData record.StructuredData) (record.Data, error) {
+	var key record.Data
+	ky, ok := structData["key"]
+	if !ok {
+		return key, cerrors.Errorf("record payload after doesn't contain key")
+	}
+
+	switch k := ky.(type) {
+	case record.Data:
+		key = k
+	case string:
+		key = record.RawData{Raw: []byte(k)}
+	default:
+		return key, cerrors.Errorf("expected a record.Data or a string, got %T", k)
+	}
+
+	return key, nil
+}
+
+// UnwrapPayload extracts payload from a structuredData record.
+func (o *openCDCUnwrapper) UnwrapPayload(structData record.StructuredData) (record.Change, error) {
+	var payload record.Change
+	pl, ok := structData["payload"]
+	if !ok {
+		return payload, cerrors.Errorf("record payload doesn't contain payload")
+	}
+
+	switch p := pl.(type) {
+	case record.Change:
+		payload = p
+	case map[string]interface{}:
+		afterData, ok := p["after"]
+		if !ok {
+			return payload, cerrors.Errorf("record payload after doesn't contain payload.after")
+		}
+
+		data, ok := afterData.(map[string]interface{})
+		if !ok {
+			return payload, cerrors.Errorf("record payload after payload.after is not a map")
+		}
+
+		convertedData := make(record.StructuredData, len(data))
+		for k, v := range data {
+			convertedData[k] = v
+		}
+
+		payload = record.Change{
+			Before: nil,
+			After:  convertedData,
+		}
+	default:
+		return payload, cerrors.Errorf("expected a record.Change or a map[string]interface{}, got %T", p)
+	}
+	return payload, nil
+}
+
+// Unwrap replaces the whole record.payload with record.payload.after.payload except position.
+func (o *openCDCUnwrapper) Unwrap(rec record.Record) (record.Record, error) {
+	var structData record.StructuredData
+	data := rec.Payload.After
+	switch d := data.(type) {
+	case record.RawData:
+		// unmarshal raw data to structured
+		if err := json.Unmarshal(data.Bytes(), &structData); err != nil {
+			return record.Record{}, cerrors.Errorf("failed to unmarshal raw data as JSON: %w", unwrapProcType, err)
+		}
+	case record.StructuredData:
+		structData = d
+	default:
+		return record.Record{}, cerrors.Errorf("unexpected data type %T", unwrapProcType, data)
+	}
+
+	operation, err := o.UnwrapOperation(structData)
+	if err != nil {
+		return record.Record{}, err
+	}
+
+	metadata, err := o.UnwrapMetadata(structData)
+	if err != nil {
+		return record.Record{}, err
+	}
+
+	key, err := o.UnwrapKey(structData)
+	if err != nil {
+		return record.Record{}, err
+	}
+
+	payload, err := o.UnwrapPayload(structData)
+	if err != nil {
+		return record.Record{}, err
+	}
+
+	return record.Record{
+		Key:       key,
+		Position:  rec.Position,
+		Metadata:  metadata,
+		Payload:   payload,
+		Operation: operation,
+	}, nil
```

Comment:
Nitpick: it would be good to document why we decided to unwrap everything except the position.

Reply:
@hariso would something make sense to you?

`Position is the only key we preserve from the original record to mantain the reference respect other messages that will be coming from`.

Reply:
Sounds good... I'd also add the context in which we assume this is happening, and that is "chaining" pipelines.